frontend/src/main/scala/bloop/dap/DebuggeeRunner.scala

```diff
@@ -0,0 +1,95 @@
+package bloop.dap
+
+import bloop.data.{Platform, Project}
+import bloop.engine.State
+import bloop.engine.tasks.{RunMode, Tasks}
+import bloop.exec.JavaEnv
+import bloop.testing.{LoggingEventHandler, TestInternals, TestSuiteEventHandler}
+import ch.epfl.scala.bsp.ScalaMainClass
+import monix.eval.Task
+
+trait DebuggeeRunner {
+  def run(logger: DebugSessionLogger): Task[Unit]
+}
+
+private final class MainClassDebugAdapter(
+    project: Project,
+    mainClass: ScalaMainClass,
+    env: JavaEnv,
+    state: State
+) extends DebuggeeRunner {
+  def run(debugLogger: DebugSessionLogger): Task[Unit] = {
+    val workingDir = state.commonOptions.workingPath
+    val runState = Tasks.runJVM(
+      state.copy(logger = debugLogger),
+      project,
+      env,
+      workingDir,
+      mainClass.`class`,
+      mainClass.arguments.toArray,
+      skipJargs = false,
+      RunMode.Debug
+    )
+
+    runState.map(_ => ())
+  }
+}
+
+private final class TestSuiteDebugAdapter(
+    projects: Seq[Project],
+    filters: List[String],
+    state: State
+) extends DebuggeeRunner {
+  def run(debugLogger: DebugSessionLogger): Task[Unit] = {
+    val filter = TestInternals.parseFilters(filters)
+    val handler = new LoggingEventHandler(state.logger)
+
+    val sequentialTests = projects.map { project =>
+      val debugState = state.copy(logger = debugLogger)
+      Tasks.test(
+        state,
```

Comment:
ah, the perils of extracting variables ;)
```suggestion
        debugState,
```